oci-unit-tests/grafana_test.sh

```diff
@@ -0,0 +1,107 @@
+# shellcheck shell=dash
+
+# shellcheck disable=SC1090
+. "$(dirname "$0")/helper/test_helper.sh"
+
+# cheat sheet:
+#  assertTrue $?
+#  assertEquals 1 2
+#  oneTimeSetUp()
+#  oneTimeTearDown()
+#  setUp() - run before each test
+#  tearDown() - run after each test
+
+# The name of the temporary docker network we will create for the
+# tests.
+readonly DOCKER_PREFIX=oci_grafana_test
+readonly DOCKER_IMAGE="squeakywheel/grafana:edge"
+readonly LOCAL_PORT=63180
+
+oneTimeSetUp() {
+    # Cleanup stale resources
+    tearDown
+}
+
+tearDown() {
+    for container in $(docker ps --filter "name=$DOCKER_PREFIX" --format "{{.Names}}"); do
+        debug "Removing container ${container}"
+        stop_container_sync "${container}"
+    done
+}
+
+docker_run_server() {
+    suffix=$(tr -dc 'a-zA-Z0-9' < /dev/urandom | fold -w 8 | head -n 1)
+    docker run \
+       --rm \
+       -d \
+       --name "${DOCKER_PREFIX}_${suffix}" \
+       "$@" \
+       $DOCKER_IMAGE
+}
+
+wait_grafana_container_ready() {
+    local container="${1}"
+    local log="HTTP Server Listen"
+    wait_container_ready "${container}" "${log}"
+}
+
+test_default_config() {
+    debug "Creating all-defaults Grafana container"
+    container=$(docker_run_server -p "$LOCAL_PORT:3000")
+
+    assertNotNull "Failed to start the container" "${container}" || return 1
+    wait_grafana_container_ready "${container}" || return 1
+
+    docker exec "$container" pgrep grafana-server > /dev/null
+    assertTrue $?
+
+    # Without Javascript the web UI doesn't do much. 
+    curl -Ss http://127.0.0.1:$LOCAL_PORT/login | grep -Fq "<title>Grafana</title>"
+    assertTrue $?
+
+    curl -Ss "http://[::1]:$LOCAL_PORT/login" | grep -Fq "<title>Grafana</title>"
+    assertTrue $?
+}
+
+test_default_config_ipv6() {
+    debug "Creating all-defaults Grafana container"
+    container=$(docker_run_server -p "$LOCAL_PORT:3000")
+
+    assertNotNull "Failed to start the container" "${container}" || return 1
+    wait_grafana_container_ready "${container}" || return 1
+
+    curl -Ss "http://[::1]:$LOCAL_PORT/login" | grep -Fq "<title>Grafana</title>"
+    assertTrue $?
+}
+
+test_persistent_storage() {
+    grafana_scratch=$(mktemp -d /tmp/grafana-XXXXXXXX)
+
+    debug "Creating Grafana container with persistent storage"
+    uid=$(id -u)
+    container=$(docker_run_server --user "$uid" -p "$LOCAL_PORT:3000" -v "$grafana_scratch:/var/lib/grafana")
+
+    assertNotNull "Failed to start the container" "${container}" || return 1
+    wait_grafana_container_ready "${container}" || return 1
+
+    curl -Ss http://127.0.0.1:$LOCAL_PORT/login | grep -Fq "<title>Grafana</title>"
+    assertTrue $?
+
+    debug "Shutting down container"
+    tearDown
```

Comment:
I think it's better to call `stop_container_sync "${container}"` here, because the `tearDown` function is supposed to be called only when the test finishes (and it may do more than shutdown the container).

Reply:
+1